metadata/src/main/java/org/apache/kafka/migration/KRaftMigrationDriver.java

```diff
@@ -0,0 +1,414 @@
+        }
+    }
+
+    class BrokerIdChangeEvent implements EventQueue.Event {
+        private final int brokerId;
+
+        BrokerIdChangeEvent(int brokerId) {
+            this.brokerId = brokerId;
+        }
+
+        @Override
+        public void run() throws Exception {
+            // TODO not sure this is expected. Can registration data change at runtime?
+            log.debug("Broker {} changed. New registration: {}", brokerId, client.readBrokerRegistration(brokerId));
+        }
+
+        @Override
+        public void handleException(Throwable e) {
+            log.error("Had an exception in " + this.getClass().getSimpleName(), e);
+        }
+    }
+
+    class KRaftLeaderEvent implements EventQueue.Event {
+        private final boolean isActive;
+        private final int kraftControllerId;
+        private final int kraftControllerEpoch;
+
+        KRaftLeaderEvent(boolean isActive, int kraftControllerId, int kraftControllerEpoch) {
+            this.isActive = isActive;
+            this.kraftControllerId = kraftControllerId;
+            this.kraftControllerEpoch = kraftControllerEpoch;
+        }
+
+        @Override
+        public void run() throws Exception {
+            if (migrationState == MigrationState.UNINITIALIZED) {
+                // If we get notified about being the active controller before we have initialized, we need
+                // to reschedule this event.
+                eventQueue.append(new PollEvent());
+                eventQueue.append(this);
+                return;
+            }
+
+            if (!isActive) {
+                apply("KRaftLeaderEvent is active", state -> state.mergeWithControllerState(ZkControllerState.EMPTY));
+                transitionTo(MigrationState.INACTIVE);
+            } else {
+                // Apply the new KRaft state
+                apply("KRaftLeaderEvent not active", state -> state.withNewKRaftController(kraftControllerId, kraftControllerEpoch));
+                // Instead of doing the ZK write directly, schedule as an event so that we can easily retry ZK failures
+                transitionTo(MigrationState.NEW_LEADER);
+                eventQueue.append(new BecomeZkLeaderEvent());
+            }
+        }
+
+        @Override
+        public void handleException(Throwable e) {
+            log.error("Had an exception in " + this.getClass().getSimpleName(), e);
+        }
+    }
+
+    class BecomeZkLeaderEvent extends ZkWriteEvent {
+        @Override
+        public void run() throws Exception {
+            ZkControllerState zkControllerState = client.claimControllerLeadership(
+                recoveryState.kraftControllerId(), recoveryState.kraftControllerEpoch());
+            apply("BecomeZkLeaderEvent", state -> state.mergeWithControllerState(zkControllerState));
+
+            if (!recoveryState.zkMigrationComplete()) {
+                transitionTo(MigrationState.ZK_MIGRATION);
```

Comment:
I think BECOME_LEADER should move to NOT_READY. Based on the heartbeats, the Hearbeats event should transition the state to either NOT_READY or ZK_MIGRATION.